src/Bicep.Core/Emit/ExpressionConverter.cs

```diff
@@ -207,292 +248,361 @@ private LanguageExpression ConvertArrayAccess(ArrayAccessSyntax arrayAccess)
-
-                    return this.CreateConverterForIndexReplacement(movedSyntax, arrayAccess.IndexExpression, arrayAccess)
-                        .GetReferenceExpression(resource, arrayAccess.IndexExpression, true);
+                    var indexContent = TryGetReplacementContext(resource, arrayAccess.IndexExpression, arrayAccess);
+                    return GetResourceReference(resource, indexContent, full: true);
                 }
 
-                switch (this.context.SemanticModel.GetSymbolInfo(arrayAccess.BaseExpression))
+                if (context.SemanticModel.GetSymbolInfo(arrayAccess.BaseExpression) is ModuleSymbol { IsCollection: true } moduleSymbol)
                 {
-                    case ModuleSymbol { IsCollection: true } moduleSymbol:
-                        var moduleConverter = this.CreateConverterForIndexReplacement(ExpressionConverter.GetModuleNameSyntax(moduleSymbol), arrayAccess.IndexExpression, arrayAccess);
-
-                        // TODO: Can this return a language expression?
-                        return moduleConverter.ToFunctionExpression(arrayAccess.BaseExpression);
+                    var indexContent = TryGetReplacementContext(ExpressionConverter.GetModuleNameSyntax(moduleSymbol), arrayAccess.IndexExpression, arrayAccess);
+                    return GetModuleOutputsReference(moduleSymbol, indexContent);
                 }
             }
 
-            return AppendProperties(
-                ToFunctionExpression(arrayAccess.BaseExpression),
-                ConvertExpression(arrayAccess.IndexExpression));
+            return new ArrayAccessOperation(
+                ConvertExpressionOperation(arrayAccess.BaseExpression),
+                ConvertExpressionOperation(arrayAccess.IndexExpression));
         }
 
-        private LanguageExpression ConvertResourcePropertyAccess(ResourceMetadata resource, SyntaxBase? indexExpression, string propertyName)
+        public LanguageExpression ConvertOperation(Operation operation)
         {
-            if (!resource.IsAzResource)
+            switch (operation)
             {
-                // For an extensible resource, always generate a 'reference' statement.
-                // User-defined properties appear inside "properties", so use a non-full reference.
-                return AppendProperties(
-                    GetReferenceExpression(resource, indexExpression, false),
-                    new JTokenExpression(propertyName));
-            }
+                case ConstantValueOperation op:
+                    {
+                        return op.Value switch
+                        {
+                            string value => new JTokenExpression(value),
+                            long value when value <= int.MaxValue && value >= int.MinValue => new JTokenExpression((int)value),
+                            long value => CreateFunction("json", new JTokenExpression(value.ToInvariantString())),
```

Comment:
Do we need to make a distinction between values that will fit in an int32 vs others? ARM seems to handle numeric literals > `int.MaxValue`.